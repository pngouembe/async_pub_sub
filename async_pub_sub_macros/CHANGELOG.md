# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.4](https://github.com/pngouembe/async_pub_sub/compare/async_pub_sub_macros-v0.1.3...async_pub_sub_macros-v0.1.4) - 2025-07-05

### Added

- forbid the use of reference in rpc interfaces

### Fixed

- remove license file in the cargo.toml

### Other

- bump the crate versions
