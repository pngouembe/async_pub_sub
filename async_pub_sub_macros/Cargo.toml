[package]
name = "async_pub_sub_macros"
version = "0.1.0"
edition = "2021"

[lib]
proc-macro = true

[dependencies]
syn = { version = "2.0", features = ["full", "extra-traits"] }
quote = "1.0"
proc-macro2 = "1.0"
heck = "0.4"
futures = "0.3.31"

[dev-dependencies]
macrotest = "1.1.0"
trybuild = "1.0"
tokio-test = "0.4"
tokio = { version = "1.0", features = ["full"] }
async_pub_sub = { path = "../async_pub_sub" }
