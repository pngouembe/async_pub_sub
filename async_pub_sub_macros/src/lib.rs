//! Asynchronous publish-subscribe macro library for Rust.
#![doc = include_str!("../README.md")]

mod helpers;
mod publisher;
mod route;
mod rpc;
mod subscriber;

use proc_macro::TokenStream;
use syn::parse_macro_input;

/// Automatically implements the `Subscriber` trait for a struct.
///
/// This macro can be used in three ways:
/// 1. For a single subscriber field - implements the `Subscriber` trait directly
/// 2. For multiple subscriber fields - implements the `SubscriberWrapper` trait for each field
/// 3. For fields marked with the `#[subscriber(type)]` attribute - specifies the message type
///
/// # Examples
/// ```rust
/// use async_pub_sub::Subscriber;
/// use async_pub_sub::SubscriberImpl;
/// use async_pub_sub_macros::DeriveSubscriber;
///
/// // Single subscriber
/// #[derive(DeriveSubscriber)]
/// struct MySubscriber<S: Subscriber> {
///     subscriber: S,
/// }
///
/// // Multiple subscribers
/// #[derive(DeriveSubscriber)]
/// struct MultiSubscriber<A, B> where
///     A: Subscriber<Message = i32>,
///     B: Subscriber<Message = String>,
/// {
///     subscriber_a: A,
///     subscriber_b: B,
/// }
///
/// // Using attribute to specify message type
/// #[derive(DeriveSubscriber)]
/// struct TypedSubscriber {
///     #[subscriber(i32)]
///     subscriber: SubscriberImpl<i32>,
/// }
/// ```
#[proc_macro_derive(DeriveSubscriber, attributes(subscriber))]
pub fn derive_subscriber(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as syn::DeriveInput);
    subscriber::derive_subscriber_impl(input)
}

/// Automatically implements the `Publisher` trait for a struct.
///
/// This macro can be used in three ways:
/// 1. For a single publisher field - implements the `Publisher` trait directly
/// 2. For multiple publisher fields - implements the `PublisherWrapper` trait for each field
/// 3. For fields marked with the `#[publisher(type)]` attribute - specifies the message type
///
/// # Examples
/// ```rust
/// use async_pub_sub::{Publisher, PublisherImpl};
/// use async_pub_sub_macros::DerivePublisher;
///
/// // Using a generic publisher
/// #[derive(DerivePublisher)]
/// struct MyPublisher<P: Publisher> {
///     publisher: P,
/// }
///
/// // Multiple publishers
/// #[derive(DerivePublisher)]
/// struct MultiPublisher<A, B> where
///     A: Publisher<Message = i32>,
///     B: Publisher<Message = String>,
/// {
///     publisher_a: A,
///     publisher_b: B,
/// }
///
/// // Using attribute to specify message type
/// #[derive(DerivePublisher)]
/// struct TypedPublisher {
///     #[publisher(i32)]
///     publisher: PublisherImpl<i32>,
/// }
/// ```
#[proc_macro_derive(DerivePublisher, attributes(publisher))]
pub fn derive_publisher(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as syn::DeriveInput);
    publisher::derive_publisher_impl(input)
}

/// Generates the necessary code for defining RPC interfaces.
///
/// This macro generates:
/// - Message enums for RPC communication (with optional derive attributes)
/// - Client traits
/// - Server traits
///
/// # Examples
/// ```rust
/// use async_pub_sub_macros::rpc_interface;
///
/// // Basic usage
/// #[rpc_interface]
/// trait MyRpcInterface {
///     async fn my_method(&self, arg: i32) -> String;
/// }
///
/// // With derive attributes for the generated message enum
/// #[rpc_interface(Debug)]
/// trait MyLoggableRpcInterface {
///     async fn get_data(&self) -> String;
///     async fn set_data(&mut self, data: String);
/// }
/// ```
///
/// The derive attributes provided to the macro will be applied to the generated
/// message enum. Common derive attributes.
/// As an example, you might want to add the `Debug` derive attribute to the
/// generated message enum to enable debugging the communication between you client and you server.
#[proc_macro_attribute]
pub fn rpc_interface(attr: TokenStream, item: TokenStream) -> TokenStream {
    let input = parse_macro_input!(item as syn::Item);
    rpc::generate_rpc_interface(attr, input)
}

/// Creates a connection between a single publisher and subscriber.
///
/// # Examples
/// ```rust
/// use async_pub_sub_macros::route;
/// use async_pub_sub::{PublisherImpl, SubscriberImpl};
///
/// // Define your publisher and subscriber
/// let mut publisher = PublisherImpl::new("publisher", 1);
/// let mut subscriber = SubscriberImpl::new("subscriber");
///
/// route!(publisher -> subscriber: i32).unwrap();
/// ```
#[proc_macro]
pub fn route(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as route::RouteInput);
    route::generate_route(input)
}

/// Creates connections between multiple publishers and subscribers.
///
/// # Examples
/// ```rust
/// use async_pub_sub_macros::routes;
/// use async_pub_sub::{PublisherImpl, SubscriberImpl};
///
/// // Define your publishers and subscribers
/// let mut publisher_a = PublisherImpl::<i32>::new("publisher_a", 1);
/// let mut publisher_b = PublisherImpl::new("publisher_b", 1);
/// let mut subscriber_a = SubscriberImpl::new("subscriber_a");
/// let mut subscriber_b = SubscriberImpl::new("subscriber_b");
///
/// routes! {
///     publisher_a -> subscriber_a,
///     publisher_b -> subscriber_b: String,
/// }.unwrap();
/// ```
#[proc_macro]
pub fn routes(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as route::RoutesInput);
    route::generate_routes(input)
}
