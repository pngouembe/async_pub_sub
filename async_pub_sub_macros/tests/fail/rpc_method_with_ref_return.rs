use async_pub_sub_macros::rpc_interface;

#[rpc_interface(Clone, Debug)]
trait TestRpc {
    async fn method_with_ref_return(&self) -> &String;
    async fn another_method(&self, arg1: u32) -> u32;
}

fn main() {}