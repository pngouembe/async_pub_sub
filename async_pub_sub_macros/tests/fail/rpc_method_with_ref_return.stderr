error: References in RPC method return types are not supported yet. Method 'method_with_ref_return' returns a reference (& String). Please use owned types.
 --> tests/fail/rpc_method_with_ref_return.rs:5:47
  |
5 |     async fn method_with_ref_return(&self) -> &String;
  |                                               ^^^^^^^
